perf: Suppress telemetry using ContextFlags(usize) instead of bool

bantonsson opened the pull request:
## Changes

This PR aligns the `Context` struct and changes a `bool` into a flag field. It tries to mitigate the performance impact of #2821 on context attach/detach operations.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

scottgerring commented:
👷 build bot output from [this run](https://github.com/open-telemetry/opentelemetry-rust/actions/runs/14082168546/job/39437405545?pr=2861): 

| name                                                  | baseDuration   | changesDuration   | difference |
|-------------------------------------------------------|---------------|------------------|------------|
| context_attach/nested_cx/empty_cx                    | 31.7±0.14ns   | **19.1±0.27ns**  | -40        |
| context_attach/nested_cx/single_value_cx             | 33.7±0.22ns   | **19.4±0.44ns**  | -42        |
| context_attach/nested_cx/span_cx                     | 33.0±0.29ns   | **19.4±0.30ns**  | -41        |
| context_attach/out_of_order_cx_drop/empty_cx         | 38.7±0.18ns   | **19.8±0.19ns**  | -49        |
| context_attach/out_of_order_cx_drop/single_value_cx  | 40.1±0.13ns   | **20.8±0.48ns**  | -48        |
| context_attach/out_of_order_cx_drop/span_cx          | 39.7±0.22ns   | **20.8±0.32ns**  | -48        |
| context_attach/single_cx/empty_cx                    | 18.1±0.22ns   | **12.5±0.31ns**  | -31        |
| context_attach/single_cx/single_value_cx             | 17.6±0.08ns   | **12.5±0.13ns**  | -29        |
| context_attach/single_cx/span_cx                     | 17.3±0.13ns   | **12.5±0.25ns**  | -28        |

bantonsson commented:
These are the benchmark numbers from [this run](https://github.com/open-telemetry/opentelemetry-rust/actions/runs/14101769547/job/39499330671?pr=2861):

And yes, the suppression check is slower, but on the other hand context attach and cloning is way faster.

| name                                                          | baseDuration         | changesDuration      | difference |
|---------------------------------------------------------------|----------------------|----------------------|------------|
| `context/has_active_span/in-cx/alt`                           | `8.4±0.01ns`         | `8.4±0.02ns`         | `0.0`      |
| `context/has_active_span/in-cx/spec`                          | `4.4±0.06ns`         | `4.4±0.07ns`         | `0.0`      |
| `context/has_active_span/no-cx/alt`                           | `8.4±0.02ns`         | `8.4±0.01ns`         | `0.0`      |
| `context/has_active_span/no-cx/spec`                          | `4.4±0.09ns`         | `4.4±0.05ns`         | `0.0`      |
| `context/has_active_span/no-sdk/alt`                          | `8.4±0.03ns`         | `8.4±0.01ns`         | `0.0`      |
| `context/has_active_span/no-sdk/spec`                         | `4.4±0.10ns`         | `4.4±0.06ns`         | `0.0`      |
| `context/is_recording/in-cx/alt`                              | `4.7±0.06ns`         | `4.7±0.06ns`         | `0.0`      |
| `context/is_recording/in-cx/spec`                             | `6.6±0.08ns`         | `6.6±0.07ns`         | `0.0`      |
| `context/is_recording/no-cx/alt`                              | `4.7±0.05ns`         | `4.7±0.04ns`         | `0.0`      |
| `context/is_recording/no-cx/spec`                             | `6.5±0.06ns`         | `6.6±0.15ns`         | `0.0`      |
| `context/is_recording/no-sdk/alt`                             | `4.7±0.07ns`         | `4.7±0.07ns`         | `0.0`      |
| `context/is_recording/no-sdk/spec`                            | `6.5±0.07ns`         | `6.5±0.09ns`         | `0.0`      |
| `context/is_sampled/in-cx/alt`                                | `8.7±0.02ns`         | `8.7±0.03ns`         | `0.0`      |
| `context/is_sampled/in-cx/spec`                               | `4.6±0.06ns`         | `4.6±0.07ns`         | `0.0`      |
| `context/is_sampled/no-cx/alt`                                | `8.7±0.04ns`         | `8.7±0.02ns`         | `0.0`      |
| `context/is_sampled/no-cx/spec`                               | `4.7±0.06ns`         | `4.7±0.09ns`         | `0.0`      |
| `context/is_sampled/no-sdk/alt`                               | `8.7±0.03ns`         | `8.7±0.03ns`         | `0.0`      |
| `context/is_sampled/no-sdk/spec`                              | `4.7±0.05ns`         | `4.7±0.07ns`         | `0.0`      |
| `context_attach/nested_cx/empty_cx`                           | `44.6±0.31ns`        | **`19.2±0.80ns`**    | `-57`      |
| `context_attach/nested_cx/single_value_cx`                    | `44.6±0.30ns`        | **`19.6±0.42ns`**    | `-56`      |
| `context_attach/nested_cx/span_cx`                            | `44.6±0.44ns`        | **`19.5±0.20ns`**    | `-56`      |
| `context_attach/out_of_order_cx_drop/empty_cx`                | `38.5±0.40ns`        | **`20.0±0.41ns`**    | `-48`      |
| `context_attach/out_of_order_cx_drop/single_value_cx`         | `38.6±0.43ns`        | **`21.2±0.42ns`**    | `-45`      |
| `context_attach/out_of_order_cx_drop/span_cx`                 | `38.6±0.45ns`        | **`21.1±0.41ns`**    | `-45`      |
| `context_attach/single_cx/empty_cx`                           | `22.8±0.20ns`        | **`10.3±0.19ns`**    | `-55`      |
| `context_attach/single_cx/single_value_cx`                    | `22.0±0.22ns`        | **`11.0±0.20ns`**    | `-50`      |
| `context_attach/single_cx/span_cx`                            | `22.0±0.25ns`        | **`11.0±0.11ns`**    | `-50`      |
| `telemetry_suppression/enter_telemetry_suppressed_scope`      | `28.4±0.09ns`        | **`24.5±0.96ns`**    | `-14`      |
| `telemetry_suppression/is_current_telemetry_suppressed_false` | **`1.3±0.02ns`**     | `1.9±0.03ns`         | `+47`      |
| `telemetry_suppression/is_current_telemetry_suppressed_true`  | **`1.3±0.01ns`**     | `2.1±0.04ns`         | `+68`      |
| `telemetry_suppression/normal_attach`                         | `28.5±0.16ns`        | **`10.0±0.18ns`**    | `-65`      |

bantonsson commented:
This is still a draft until #2870 has been merged and all benchmarks are run properly.

bantonsson commented:
New performance numbers from new approach in [this run](https://github.com/open-telemetry/opentelemetry-rust/actions/runs/14245896888/job/39926647144?pr=2861):

| name                                                          | baseDuration         | changesDuration      | difference |
|---------------------------------------------------------------|----------------------|----------------------|------------|
| `context/has_active_span/in-cx/alt`                           | `8.4±0.03ns`         | `8.4±0.05ns`         | `0.0`      |
| `context/has_active_span/in-cx/spec`                          | `5.2±0.17ns`         | **`5.0±0.16ns`**     | `-2.9`     |
| `context/has_active_span/no-cx/alt`                           | `8.4±0.03ns`         | `8.4±0.03ns`         | `0.0`      |
| `context/has_active_span/no-cx/spec`                          | `5.0±0.17ns`         | **`4.7±0.19ns`**     | `-6.5`     |
| `context/has_active_span/no-sdk/alt`                          | `8.4±0.02ns`         | `8.4±0.02ns`         | `0.0`      |
| `context/has_active_span/no-sdk/spec`                         | `5.0±0.18ns`         | **`4.7±0.17ns`**     | `-6.5`     |
| `context/is_recording/in-cx/alt`                              | `4.7±0.19ns`         | `4.7±0.14ns`         | `0.0`      |
| `context/is_recording/in-cx/spec`                             | `7.5±0.22ns`         | `7.5±0.23ns`         | `0.0`      |
| `context/is_recording/no-cx/alt`                              | `4.7±0.14ns`         | `4.7±0.17ns`         | `0.0`      |
| `context/is_recording/no-cx/spec`                             | **`7.2±0.36ns`**     | `7.2±0.39ns`         | `+1.0`     |
| `context/is_recording/no-sdk/alt`                             | `4.7±0.13ns`         | `4.7±0.14ns`         | `0.0`      |
| `context/is_recording/no-sdk/spec`                            | `7.2±0.21ns`         | `7.2±0.24ns`         | `0.0`      |
| `context/is_sampled/in-cx/alt`                                | `8.7±0.03ns`         | `8.7±0.04ns`         | `0.0`      |
| `context/is_sampled/in-cx/spec`                               | `5.4±0.15ns`         | **`5.3±0.18ns`**     | `-0.99`    |
| `context/is_sampled/no-cx/alt`                                | `8.7±0.03ns`         | `8.7±0.04ns`         | `0.0`      |
| `context/is_sampled/no-cx/spec`                               | `5.1±0.53ns`         | **`5.0±0.17ns`**     | `-0.99`    |
| `context/is_sampled/no-sdk/alt`                               | `8.7±0.03ns`         | `8.7±0.05ns`         | `0.0`      |
| `context/is_sampled/no-sdk/spec`                              | `5.0±0.06ns`         | `5.0±0.19ns`         | `0.0`      |
| `context_attach/nested_cx/empty_cx`                           | `47.4±1.16ns`        | **`39.0±1.06ns`**    | `-18`      |
| `context_attach/nested_cx/single_value_cx`                    | `48.7±1.19ns`        | **`42.8±1.00ns`**    | `-12`      |
| `context_attach/nested_cx/span_cx`                            | `48.5±0.44ns`        | **`43.0±4.56ns`**    | `-12`      |
| `context_attach/out_of_order_cx_drop/empty_cx`                | `41.7±0.94ns`        | **`40.6±1.20ns`**    | `-2.9`     |
| `context_attach/out_of_order_cx_drop/single_value_cx`         | `42.3±2.58ns`        | **`42.0±0.89ns`**    | `-0.99`    |
| `context_attach/out_of_order_cx_drop/span_cx`                 | `42.5±0.87ns`        | **`42.0±0.85ns`**    | `-0.99`    |
| `context_attach/single_cx/empty_cx`                           | `24.0±0.59ns`        | **`19.4±0.36ns`**    | `-19`      |
| `context_attach/single_cx/single_value_cx`                    | `23.4±0.52ns`        | `23.4±0.44ns`        | `0.0`      |
| `context_attach/single_cx/span_cx`                            | `23.4±0.65ns`        | **`23.1±0.51ns`**    | `-2.0`     |
| `exporter_disabled_concurrent_processor`                      | **`2.9±0.29ns`**     | `3.1±0.08ns`         | `+7.0`     |
| `exporter_disabled_simple_processor`                          | **`6.0±0.26ns`**     | `6.2±0.18ns`         | `+5.0`     |
| `telemetry_suppression/enter_telemetry_suppressed_scope`      | `27.1±0.18ns`        | **`25.2±0.46ns`**    | `-7.4`     |
| `telemetry_suppression/is_current_telemetry_suppressed_false` | `1.4±0.02ns`         | **`1.3±0.03ns`**     | `-6.5`     |
| `telemetry_suppression/is_current_telemetry_suppressed_true`  | `1.4±0.05ns`         | **`1.3±0.03ns`**     | `-6.5`     |
| `telemetry_suppression/normal_attach`                         | `30.4±0.62ns`        | **`28.0±0.86ns`**    | `-7.4`     |

bantonsson commented:
@cijothomas These are the numbers from my `M1 Max` Laptop (where I see a slight regression in checks and large improvements in entering)
```
Benchmarking telemetry_suppression/enter_telemetry_suppressed_scope: Collecting 100 samples in estimated 2.0001 s (135M itelemetry_suppression/enter_telemetry_suppressed_scope
                        time:   [14.479 ns 14.511 ns 14.549 ns]
                        change: [-21.316% -20.968% -20.611%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) low mild
  2 (2.00%) high mild
  7 (7.00%) high severe
telemetry_suppression/normal_attach
                        time:   [15.431 ns 15.483 ns 15.550 ns]
                        change: [-18.631% -18.300% -17.951%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  6 (6.00%) low mild
  1 (1.00%) high mild
  5 (5.00%) high severe
Benchmarking telemetry_suppression/is_current_telemetry_suppressed_false: Collecting 100 samples in estimated 2.0000 s (1telemetry_suppression/is_current_telemetry_suppressed_false
                        time:   [1.1357 ns 1.1436 ns 1.1545 ns]
                        change: [+2.1229% +2.7044% +3.3306%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
Benchmarking telemetry_suppression/is_current_telemetry_suppressed_true: Collecting 100 samples in estimated 2.0000 s (1.telemetry_suppression/is_current_telemetry_suppressed_true
                        time:   [1.1311 ns 1.1367 ns 1.1438 ns]
                        change: [+1.1998% +1.6296% +2.0577%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 23 outliers among 100 measurements (23.00%)
  2 (2.00%) low severe
  3 (3.00%) low mild
  4 (4.00%) high mild
  14 (14.00%) high severe
```
And these are the numbers from my `AMD Ryzen 5 3600 2.2GHz` box, where I see a slight improvement in checks and large improvements in entering).
```
Benchmarking telemetry_suppression/enter_telemetry_suppressed_scope: Collecting 100 samples in estimated 2.0001 s (85M iteratio
telemetry_suppression/enter_telemetry_suppressed_scope
                        time:   [23.078 ns 23.081 ns 23.083 ns]
                        change: [-16.623% -16.588% -16.558%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
telemetry_suppression/normal_attach
                        time:   [24.008 ns 24.024 ns 24.039 ns]
                        change: [-17.823% -17.631% -17.412%] (p = 0.00 < 0.05)
                        Performance has improved.
Benchmarking telemetry_suppression/is_current_telemetry_suppressed_false: Collecting 100 samples in estimated 2.0000 s (2.8B it
telemetry_suppression/is_current_telemetry_suppressed_false
                        time:   [715.65 ps 715.73 ps 715.82 ps]
                        change: [-1.7774% -1.7447% -1.7127%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  3 (3.00%) high severe
Benchmarking telemetry_suppression/is_current_telemetry_suppressed_true: Collecting 100 samples in estimated 2.0000 s (2.8B ite
telemetry_suppression/is_current_telemetry_suppressed_true
                        time:   [715.54 ps 715.63 ps 715.73 ps]
                        change: [-3.6408% -3.0627% -2.5393%] (p = 0.00 < 0.05)
                        Performance has improved.
```

bantonsson commented:
The [benchmark numbers comment](https://github.com/open-telemetry/opentelemetry-rust/pull/2861#issuecomment-2777828099) have been updated with the latest run.

bantonsson commented:
It would be great if there could be some extra manual testing of this @cijothomas @lalitb so we can come to a conclusion.

cijothomas commented:
```txt
telemetry_suppression/enter_telemetry_suppressed_scope
                        time:   [10.502 ns 10.524 ns 10.546 ns]
                        change: [+12.347% +12.742% +13.119%] (p = 0.00 < 0.05)
                        Performance has regressed.
telemetry_suppression/normal_attach
                        time:   [11.106 ns 11.133 ns 11.166 ns]
                        change: [+8.4063% +8.8465% +9.3006%] (p = 0.00 < 0.05)
                        Performance has regressed.
telemetry_suppression/is_current_telemetry_suppressed_false
                        time:   [743.58 ps 744.81 ps 746.44 ps]
                        change: [-0.6749% -0.3983% -0.1260%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 21 outliers among 100 measurements (21.00%)
  8 (8.00%) low severe
  3 (3.00%) low mild
  3 (3.00%) high mild
  7 (7.00%) high severe
telemetry_suppression/is_current_telemetry_suppressed_true
                        time:   [743.35 ps 744.08 ps 745.10 ps]
                        change: [-0.4459% -0.1688% +0.0707%] (p = 0.22 > 0.05)
                        No change in performance detected.
Found 20 outliers among 100 measurements (20.00%)
  3 (3.00%) low severe
  6 (6.00%) low mild
  5 (5.00%) high mild
  6 (6.00%) high severe
```

I'm seeing consistent ~10% regression for enter, no-change for checks.

bantonsson commented:
That is interesting @cijothomas. What hardware are you running on?

cijothomas commented:
> That is interesting @cijothomas. What hardware are you running on?

This was in m4 pro. I can try in a windows box and update back.

cijothomas commented:
@lalitb @utpilla could you also help run the benches for this to see if there is consistent improvement/regression?

bantonsson commented:
I can't seem to get consistently good benchmark numbers across the different architectures, so I'll close this work for now.

bantonsson committed:
perf: Suppress telemetry using ContextFlags(usize) instead of bool

The code seems to be highly sensitive to alignment, so use a bitfield
instead of a boolean.
